lib/chess/auth/user.ex

```diff
@@ -0,0 +1,33 @@
+defmodule Chess.Auth.User do
+  @moduledoc false
+
+  use Ecto.Schema
+  import Ecto.Changeset
+  alias Comeonin.Argon2
+  alias Chess.Repo
+
+  schema "users" do
+    field :password, :string, virtual: true
+    field :password_hash, :string
+    field :username, :string
+
+    timestamps()
+  end
+
+  @doc false
+  def changeset(struct, params \\ %{}) do
+    struct
+    |> cast(params, [:username, :password])
+    |> validate_required([:username, :password])
+    |> unsafe_validate_unique(:username, Repo)
+    |> hash_password()
+  end
+
+  defp hash_password(
+    %Ecto.Changeset{valid?: true, changes: %{password: password}} = changeset
+  ) do
+    change(changeset, Argon2.add_hash(password))
```

Comment:
I think this would normally look like this:

```elixir
import Ecto.Changeset

defp hash_password(changeset) do
  password = get_change(changeset, :password)
  if password do
    hashed_password = Argon2.add_hash(password)
    changeset
    |> put_change(:hashed_password, hashed_password)
  else
    changeset
  end
end
```

ie, you should access changes through Ecto's function and not the struct. You _can_ make it work and it'll work for probably years, but the function is the API to access the data or changes.